File: apps/api/src/runners/python/runner.py
import json, sys, time, os, subprocess, glob


def run_all_tests(time_limit_ms):
    tests = sorted(glob.glob("/tests/input*.in"))
    cases = []
    total_time = 0
    ok_count = 0
    for idx, inpath in enumerate(tests, start=1):
        outpath = inpath.replace("/input", "/output").replace(".in", ".out")
        start = time.time()
        try:
            p = subprocess.run(
                ["python3", sys.argv[1]]
                if len(sys.argv) > 1
                else ["python3", "/code/solution.py"],
                stdin=open(inpath, "rb"),
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                timeout=time_limit_ms / 1000,
            )
            elapsed = int((time.time() - start) * 1000)
            actual = p.stdout.decode().strip()
            expected = (
                open(outpath, "r").read().strip() if os.path.exists(outpath) else ""
            )
            status = (
                "OK"
                if actual == expected and p.returncode == 0
                else ("WA" if p.returncode == 0 else "RE")
            )
            if p.returncode != 0 and "Traceback" in p.stderr.decode():
                status = "RE"
            cases.append(
                {
                    "caseId": idx,
                    "status": status,
                    "timeMs": elapsed,
                    "stderr": p.stderr.decode(),
                }
            )
            total_time += elapsed
            if status == "OK":
                ok_count += 1
        except subprocess.TimeoutExpired:
            cases.append({"caseId": idx, "status": "TLE", "timeMs": time_limit_ms})
            total_time += time_limit_ms
        except Exception as e:
            cases.append({"caseId": idx, "status": "RE", "timeMs": 0, "stderr": str(e)})
    return cases, total_time, ok_count, len(tests)


if __name__ == "__main__":
    # runner receives payload on stdin for compatibility but arguments also allowed
    try:
        payload = json.load(sys.stdin)
    except Exception:
        payload = {}
    # prefer source_file from payload but also use /code/solution.py by default
    src = payload.get("source_file", "/code/solution.py")
    time_limit = int(payload.get("time_limit_ms", payload.get("time_limit", 1500)))
    # place the source path as sys.argv[1] so subprocess runs proper file
    sys.argv = [sys.argv[0], src]
    cases, total_time, ok_count, total = run_all_tests(time_limit)
    if total == ok_count:
        status = "ACCEPTED"
    elif ok_count == 0:
        status = "WRONG_ANSWER" if any(c["status"] == "WA" for c in cases) else "ERROR"
    else:
        status = "PARTIAL"
    out = {"status": status, "timeMsTotal": total_time, "cases": cases}
    print(json.dumps(out))

File: apps/api/src/runners/node/runner.js
import fs from "fs";
import { spawnSync } from "child_process";
import { glob } from "glob";

// === Leer configuración desde stdin ===
const raw = fs.readFileSync(0, "utf8");
const payload = JSON.parse(raw || "{}");

const src = payload.source_file || "/code/solution.js";
const timeLimit = payload.time_limit_ms || payload.time_limit || 1500;

// === 1️⃣ Verificar errores de sintaxis (como compilación) ===
const check = spawnSync("node", ["--check", src], { encoding: "utf8" });
if (check.status !== 0) {
    console.log(
        JSON.stringify({
            status: "COMPILATION_ERROR",
            timeMsTotal: 0,
            cases: [],
            stderr: (check.stderr || "").trim(),
        }),
    );
    process.exit(0);
}

// === 2️⃣ Buscar archivos de test ===
const inputs = glob.sync("/tests/*.in").sort();
const cases = [];
let totalTime = 0;
let correct = 0;
let runtimeErrors = 0;
let wrongAnswers = 0;

// === 3️⃣ Ejecutar cada test ===
for (let i = 0; i < inputs.length; i++) {
    const inFile = inputs[i];
    const base = inFile.match(/input(\d+)\.in$/)?.[1] || "";
    const outFile = `/tests/output${base}.out`;

    const input = fs.existsSync(inFile) ? fs.readFileSync(inFile, "utf8") : "";

    const expected = fs.existsSync(outFile)
        ? fs.readFileSync(outFile, "utf8").trim()
        : "";

    const start = Date.now();
    let proc;

    try {
        proc = spawnSync("node", [src], {
            input,
            timeout: timeLimit,
            encoding: "utf8",
            maxBuffer: 10 * 1024 * 1024,
            shell: false,
        });
    } catch (e) {
        // Error lanzado directamente por spawnSync
        cases.push({
            caseId: i + 1,
            status: "RUNTIME_ERROR",
            timeMs: 0,
            stderr: e.message || String(e),
        });
        runtimeErrors++;
        continue;
    }

    const elapsed = Date.now() - start;
    totalTime += elapsed;

    // Timeout (TLE)
    if (proc.error && proc.error.code === "ETIMEDOUT") {
        cases.push({
            caseId: i + 1,
            status: "TLE",
            timeMs: timeLimit,
            stderr: "",
        });
        runtimeErrors++;
        continue;
    }

    // Proceso con error (exit code != 0)
    if (proc.status !== 0) {
        const stderrMsg =
            (proc.stderr || "").trim() || (proc.error?.message ?? "");
        cases.push({
            caseId: i + 1,
            status: "RUNTIME_ERROR",
            timeMs: elapsed,
            stderr: stderrMsg,
        });
        runtimeErrors++;
        continue;
    }

    // Comparar salidas
    const actual = (proc.stdout || "").trim();
    const ok = actual === expected;

    cases.push({
        caseId: i + 1,
        status: ok ? "OK" : "WRONG_ANSWER",
        timeMs: elapsed,
        stderr: ok ? "" : `Expected: "${expected}", Got: "${actual}"`,
    });

    if (ok) correct++;
    else wrongAnswers++;
}

// === 4️⃣ Determinar estado general ===
let status = "ACCEPTED";
if (runtimeErrors > 0) status = "RUNTIME_ERROR";
else if (wrongAnswers > 0) status = "WRONG_ANSWER";

console.log(JSON.stringify({ status, timeMsTotal: totalTime, cases }));
